### ISSUE TITLE
assign my_dataset.map() result to my_dataset cause faiss index lost

### ISSUE THREAD

**Issue by Oaklight** [huggingface/datasets#3769]
## Describe the bug
assigning the resulted dataset to original dataset causes lost of the faiss index

## Steps to reproduce the bug
`my_dataset` is a regular loaded dataset. It's a part of a customed dataset structure
```python
self.my_dataset.add_faiss_index('embeddings')
self.my_dataset.list_indexes()
# ['embeddings']

my_dataset2 = my_dataset.map(
    lambda x: _get_nearest_examples_batch(x['text']), batch=True
)

# the unexpected result:
my_dataset2.list_indexes()
# []

self.my_dataset.list_indexes()
# ['embeddings']
```

in case something wrong with my `_get_nearest_examples_batch()`, it's like this
```python
def _get_nearest_examples_batch(self, examples, k=5):
    queries = embed(examples)
    scores_batch, retrievals_batch = self.dataset.get_nearest_examples_batch(self.faiss_column, queries, k)
    return {
        'neighbors': [batch['text'] for batch in retrievals_batch],
        'scores': scores_batch
    }
```

## Expected results
reassignment shouldn't intact the indexes

## Actual results
reassignment drops the indexes

## Environment info

- `datasets` version: 1.18.3
- Platform: Ubuntu 20.04.3 LTS
- Python version: 3.8.12
- PyArrow version: 7.0.0

**Comment by lhoestq** [comment]
Hi ! Indeed `map` is dropping the index right now, because one can create a dataset with more or fewer rows using `map` (and therefore the index might not be relevant anymore)

I guess we could check the resulting dataset length, and if the user hasn't changed the dataset size we could keep the index, what do you think ?

**Comment by cceyda** [comment]
doing `.add_column("x",x_data)` also removes the index. the new column might be irrelevant to the index so I don't think it should drop.

**Comment by lhoestq** [comment]
I agree `add_column` shouldn't drop the index indeed ! Is it something you'd like to contribute ? I think it's just a matter of copying the `self._indexes` dictionary to the output dataset
